Parse/src/main/java/com/parse/ParsePolygon.java

```diff
@@ -0,0 +1,228 @@
+ */
+package com.parse;
+
+import android.location.Criteria;
+import android.location.Location;
+import android.os.Parcel;
+import android.os.Parcelable;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+
+import java.util.Locale;
+import java.util.List;
+
+import bolts.Continuation;
+import bolts.Task;
+
+/**
+ * {@code ParsePolygon} represents a set of coordinates that may be associated with a key
+ * in a {@link ParseObject} or used as a reference point for geo queries. This allows proximity
+ * based queries on the key.
+ *
+ * Example:
+ * <pre>
+ * List<ParseGeoPoint> points = new ArrayList<ParseGeoPoint>();
+ * points.add(new ParseGeoPoint(0,0));
+ * points.add(new ParseGeoPoint(0,1));
+ * points.add(new ParseGeoPoint(1,1));
+ * points.add(new ParseGeoPoint(1,0));
+ * ParsePolygon polygon = new ParsePolygon(points);
+ * ParseObject object = new ParseObject("PlaceObject");
+ * object.put("area", polygon);
+ * object.save();
+ * </pre>
+ */
+
+public class ParsePolygon implements Parcelable {
+
+  private List<ParseGeoPoint> coordinates;
+
+  /**
+   * Creates a new polygon with the specified {@link ParseGeoPoint}.
+   *
+   * @param coords
+   *          The polygon's coordinates.
+   */
+  public ParsePolygon(List<ParseGeoPoint> coords) {
+    setCoordinates(coords);
+  }
+
+  /**
+   * Creates a copy of {@code polygon};
+   *
+   * @param polygon
+   *          The polygon to copy.
+   */
+  public ParsePolygon(ParsePolygon polygon) {
+    this(polygon.getCoordinates());
+  }
+
+  /**
+   * Creates a new point instance from a {@link Parcel} source. This is used when unparceling a
+   * ParsePolygon. Subclasses that need Parcelable behavior should provide their own
+   * {@link android.os.Parcelable.Creator} and override this constructor.
+   *
+   * @param source The recovered parcel.
+   */
+  protected ParsePolygon(Parcel source) {
+    this(source, ParseParcelDecoder.get());
+  }
+
+  /**
+   * Creates a new point instance from a {@link Parcel} using the given {@link ParseParcelDecoder}.
+   * The decoder is currently unused, but it might be in the future, plus this is the pattern we
+   * are using in parcelable classes.
+   *
+   * @param source the parcel
+   * @param decoder the decoder
+   */
+  ParsePolygon(Parcel source, ParseParcelDecoder decoder) {
+    setCoordinates(source.readArrayList(null));
+  }
+
+
+  /**
+   * Set coordinates. Valid are Array of GeoPoint, ParseGeoPoint or Location
+   * at least 3 points
+   *
+   * @param coords
+   *          The polygon's coordinates.
+   */
+  public void setCoordinates(List<ParseGeoPoint> coords) {
+    this.coordinates = ParsePolygon._validate(coords);
+  }
+
+  /**
+   * Get coordinates.
+   */
+  public List<ParseGeoPoint> getCoordinates() {
+    return coordinates;
+  }
+
+  /**
+   * Get converts coordinate to JSONArray.
+   */
+  public JSONArray coordinatesToJSONArray() throws JSONException{
```

Comment:
Could you make this package protected? We are not exposing JSONs anywhere in the SDK